gpflow/optimizers/scipy.py

```diff
@@ -99,34 +100,37 @@ def _eval(x: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
 
     @classmethod
     def callback_func(
-        cls, variables: Variables, step_callback: StepCallback
+        cls, variables: List[tf.Variable], step_callback: StepCallback
     ) -> Callable[[np.ndarray], None]:
         step = 0  # type: int
 
         def _callback(x: np.ndarray) -> None:
             nonlocal step
             values = cls.unpack_tensors(variables, x)
-            step_callback(step=step, variables=variables, values=values)
+            step_callback(step, variables, values)
             step += 1
 
         return _callback
 
     @staticmethod
-    def pack_tensors(tensors: Iterable[tf.Tensor]) -> tf.Tensor:
+    def pack_tensors(tensors: Iterable[Union[tf.Tensor, tf.Variable]]) -> tf.Tensor:
```

Comment:
the type was wrong, we actually pass in the variables, not tensors (we *only* pass in variables, but the code also works for tensors, so didn't want to restrict needlessly)

Reply:
tf.Variable is a tensor.

Reply:
are you sure artem?
```
>>> isinstance(tf.Variable(0), tf.Tensor)
False
```

Reply:
re

> didn't want to restrict needlessly

I'd say simplicity is a good argument for ... simplicity. Also a more constrained type can be made more flexible later

Reply:
tf.Variable is, unfortunately, NOT a tensor... that's why we've had so much discussion around `TensorType` in #1427 (and I would replace all these unions with `TensorType` if #1427 was merged).

Reply:
@joelberkeley-pio, Interesting. I believe it now :) Although, this was not true in `2.0.0` version.